### ISSUE TITLE
Potential issue with `unused_kernel_region_number` in rv32i/pmp.rs

### ISSUE THREAD

**Issue by jettr** [tock/tock#2727]
Looking at the [unused_kernel_region_number][0] function in rv32i/pmp.rs, it appears that we are trying to iterate through the available PMP regions backwards since we want the app regions to be before the kernel regions, but we end up mismatching the `region` from its actual array index/`number`.

It seems like the below snippet would be correct instead. If that isn't the case, what am I missing?

```
    fn unused_kernel_region_number(&self, locked_region_mask: u64) -> Option<usize> {
        for (number, region) in self.regions.iter().rev().enumerate() {
            if self.app_memory_region.contains(&number) {
                continue;
            }
            // This region exists, but is locked
            if locked_region_mask & (1 << number) > 0 {
                continue;
            }
            if region.is_none() {
                return Some(number);
            }
        }
        None
    }
```

Basically, remove the 
`let number = MAX_AVAILABLE_REGIONS_OVER_TWO - num - 1;` line and rename `num` to `number`

@alistair23 , can you show me what I am missing here?

[0]:https://github.com/tock/tock/blob/master/arch/rv32i/src/pmp.rs#L277

**Comment by alistair23** [comment]
Yeah, I think you are right. Are you seeing a failure?

**Comment by jettr** [comment]
No, I am not seeing an issue, but I was just reading through the code as I was changing things. It couldn't understand what is was trying to do.

**Comment by jettr** [comment]
Should be move forward with a fix for this?

**Comment by alistair23** [comment]
Yeah, I think so. It looks incorrect

**Comment by jettr** [comment]
I figured out what is going on here. We should be calling `enumerate` before `rev` and the confusing line is correcting for calling `rev` before `enumerate`. I upload a "fix" that doesn't change behavior, but just documents what is happening more explicitly in #2806.

See also this [rust playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1b66e17de3c22e49cb855dde854cce09)
